blocks/main_loop.py

```diff
@@ -62,37 +67,44 @@ class MainLoop(object):
 
     Parameters
     ----------
-    model : object
-        The model object. It is entirely transparent for the main loop
-        but may be used by extensions.
-    data_stream : instance of :class:`.DataStream`.
-        The data stream.
     algorithm : object
         The training algorithm.
-    log : instance of :class:`.TrainingLog`
+    data_stream : instance of :class:`.DataStream`.
+        The data stream.
+    model : :class:`.AbstractModel` instance, optional
+        The model object. It is entirely transparent for the main loop
+        but may be used by extensions.
+    log : instance of :class:`.TrainingLog`, optional
         The log. When not given, a :class:`.TrainingLog` is created.
     extensions : list of :class:`.TrainingExtension` instances
         The training extensions. Will be called in the same order as given
         here.
 
     """
-    def __init__(self, model, data_stream, algorithm,
-                 log=None, extensions=None):
-        self.model = model
-        self.data_stream = data_stream
-        self.algorithm = algorithm
-
+    def __init__(self, algorithm, data_stream,
+                 model=None, log=None, extensions=None):
         if not log:
             log = TrainingLog()
         if not extensions:
             extensions = []
+
+        self.data_stream = data_stream
+        self.algorithm = algorithm
         self.log = log
         self.extensions = extensions
 
+        self._model = model
+
         self.status._training_started = False
         self.status._epoch_started = False
 
     @property
+    def model(self):
+        if not self._model:
+            raise ValueError("no model in this main loop" + no_model_message)
```

Comment:
This should be an `AttributeError`, I think that would also allow for `hasattr(main_loop, 'model')` checks.